### ISSUE TITLE
repetitive 'importing 300 addresses into account'

### ISSUE THREAD

**Issue by raedah** [JoinMarket-Org/joinmarket#522]
Steps
1. Start with a fresh bitcoin wallet.dat to hold your watch addresses on the rpc server.
2. Launch joinmarket with a jm wallet you have already used for sometime.
3. see:  too few addresses in [(0, 1), (1, 1), (2, 1), (3, 1)]
             importing 300 addresses into account joinmarket-wallet-abc123
4. says, restart Bitcoin Core with -rescan, do that, takes forever
5. repeat steps 2,3,4 over and over. Does it ever end?

**Comment by AdamISZ** [comment]
I think we've discussed this before, right? Is https://www.reddit.com/r/joinmarket/comments/3te8lm/too_few_addresses_in_3_1_4_0/cx5frsx not solving it? (not that I'm saying it shouldn't be fixed up, just wondering if you're talking about something new or the same thing)

**Comment by raedah** [comment]
Yes, we have discussed this before. Filed so that we can look for a permanent fix that doesn't require changing the code manually.

**Comment by raedah** [comment]
On some runs, such as the first run, it doesnt say "too fee addresses". It just goes straight to the importing step.  The code that causes this is these 15 lines here: https://github.com/JoinMarket-Org/joinmarket/blob/master/joinmarket/blockchaininterface.py#L591-L607

It just imports 20 addresses for internal and external and for each mixdepth level, so 200 total.
Instead of doing that, it should be using the wallet index_cache numbers.

**Comment by chris-belcher** [comment]
How about an option in the .cfg file which controls how many addresses to import per branch ?

**Comment by raedah** [comment]
the .cfg could be done, but I think what is most important here is that the address importing recognizes the wallet index_cache

**Comment by AdamISZ** [comment]
> But most important here is that the address importing recognizes the wallet index_cache so that recoveries do not fail.

Having recently done a lot of work on this (see #574 ) I am of the opinion that it *should* indeed be the case that simply repeating over and over again until it works is enough. And this includes the case of a miscreant counterparty requesting lots of addresses. The index_cache field in the wallet file is needed for the sync to happen correctly, and it is not changed between runs of sync_wallet. (It might be a bit "in the weeds" but note this comment: https://github.com/JoinMarket-Org/joinmarket/pull/574#discussion_r67148491)

And also, I'll backtrack a little, because I'm not yet *sure* that it's impossible to get an "infinite loop" scenario. I was able to create such a scenario in testing, but only by injecting state that doesn't happen in real runs. I can't yet find a way for it to happen for an actual bot/script/yg. In my own case, I have been running for 18 months on one wallet and never had to do anything more than re-run the sync a few times, and perhaps increase the addr_req_count. Never had to rescan, either.

**Comment by raedah** [comment]
@AdamISZ Did you rename your bitcoin core wallet.dat, and try running joinmarket from scratch with no imports. We already know the correct number of addresses to import from the index_cache, so there really is no reason to ask the user to repeat command combos to unlock the secret functionality.

**Comment by raedah** [comment]
Another case that triggers this issue is having several empty addresses that creates a large gap limit. In that case, it does not matter how many times you run the command, you will be stuck in an infinite loop. The solution is the same, to manually edit the blockchaininterface.py to increase the limit, or to respect the index_cache.

**Comment by AdamISZ** [comment]
> @AdamISZ Did you rename your bitcoin core wallet.dat, and try running joinmarket from scratch with no imports.

No - and thanks for the suggestion, because that is something I can stick into test_wallet_sync as another test vector. I'm going to spend some time investigating that now. 

> Another case that triggers this issue is having several empty addresses that creates a large gap limit.

Not usually: A correct index_cache that points to say, 100 as the last used index, with a large gap between say 70 and 95, will not create a problem because the condition is: `if unused_addr_count > gaplimit AND is_index_ahead_of_cache()`.

Large gaps create problems *on recovery*, specifically because we don't have an index_cache telling us where to stop searching.

**Comment by raedah** [comment]
@Large gaps create problems on recovery, specifically because we don't have an index_cache telling us where to stop searching.

I believe that is in fact the case here, because this code isn't obeying the index_cache.

**Comment by AdamISZ** [comment]
@raedah sorry i'm lost on your final comment there; which code isn't obeying index_cache? I was just saying that on recovery there is no index_cache available (recover = no wallet.json file to use)

**Comment by raedah** [comment]
Oh, I am talking about the where the user has their wallet.json file for recovery, but does not have the bitcoin core wallet.dat file, so all the addresses need to be imported.

**Comment by raedah** [comment]
People still getting hit by this regularly.

https://www.reddit.com/r/joinmarket/comments/4p8i56/yieldgeneratoroscillator_stuck_in_an_too_few/

https://github.com/JoinMarket-Org/joinmarket/issues/583

**Comment by AdamISZ** [comment]
The resolution of the problem in that reddit thread is in line with the investigations I did. Rerunning sync_wallet (which happens when you do wallet-tool) several times *without* rescanning should fix it. Certainly bumping addr_req_count can speed it up. As I mentioned in the comments to #574 I believe there is a case for doing this under the hood, but it's tricky.

**Comment by raedah** [comment]
related issue, https://github.com/JoinMarket-Org/joinmarket/issues/478
